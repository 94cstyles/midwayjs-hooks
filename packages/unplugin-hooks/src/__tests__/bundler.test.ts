import {
  All,
  BaseTrigger,
  Decorate,
  Delete,
  FileRouter,
  Get,
  Head,
  loadApiRoutesFromFile,
  Operator,
  OperatorType,
  Options,
  Patch,
  Post,
  Put,
} from '@midwayjs/hooks-core'
import { AbstractBundlerAdapter } from '..'
import { wrap } from 'jest-snapshot-serializer-raw'

const router = new FileRouter({
  root: '/',
  source: '/',
  routes: [
    {
      baseDir: '/',
    },
  ],
})

class TestBundlerAdapter extends AbstractBundlerAdapter {
  constructor() {
    super({
      name: 'test',
      router,
    })
  }
}

it('create bundler', async () => {
  const testBundlerAdapter = new TestBundlerAdapter()
  expect(testBundlerAdapter).toBeTruthy()
  expect(testBundlerAdapter).toBeInstanceOf(AbstractBundlerAdapter)
  expect(testBundlerAdapter.getName()).toEqual('test')
})

it('generate client', async () => {
  const testBundlerAdapter = new TestBundlerAdapter()
  expect(testBundlerAdapter.generateClient([])).toEqual('')

  const apis = loadApiRoutesFromFile(
    {
      get: Decorate(Get(), async () => {}),
      post: Decorate(Post(), async () => {}),
      put: Decorate(Put(), async () => {}),
      del: Decorate(Delete(), async () => {}),
      patch: Decorate(Patch(), async () => {}),
      head: Decorate(Head(), async () => {}),
      options: Decorate(Options(), async () => {}),
      default: Decorate(All(), async () => {}),
    },
    '/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(wrap(testBundlerAdapter.generateClient(apis))).toMatchSnapshot()
})

it('with requestClient should not generate client', async () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
        })
      },
    }
  }

  const testBundlerAdapter = new TestBundlerAdapter()

  const apis = loadApiRoutesFromFile(
    {
      custom: Decorate(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(testBundlerAdapter.generateClient(apis)).toBe('')
})

it('with requestClient generate client', async () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty<BaseTrigger>(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
          requestClient: {
            fetcher: 'ws',
            client: '@midwayjs/rpc',
          },
        })
      },
    }
  }

  const testBundlerAdapter = new TestBundlerAdapter()

  const apis = loadApiRoutesFromFile(
    {
      get: Decorate(Get(), async () => {}),
      custom: Decorate(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(testBundlerAdapter.generateClient(apis)).toMatchSnapshot()
})

it('with requestClient generate multi client', async () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty<BaseTrigger>(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
          requestClient: {
            fetcher: 'cloud',
            client: 'cloud-invoker',
          },
        })
      },
    }
  }

  const testBundlerAdapter = new TestBundlerAdapter()

  const apis = loadApiRoutesFromFile(
    {
      get: Decorate(Get(), async () => {}),
      custom: Decorate(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(wrap(testBundlerAdapter.generateClient(apis))).toMatchSnapshot()
})
