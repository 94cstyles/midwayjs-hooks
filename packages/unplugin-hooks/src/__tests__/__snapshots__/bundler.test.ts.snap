// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate client 1`] = `
import { $fetch as request$0 } from '@midwayjs/fetch';

export function get(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'GET', path: '/get' },
    functionId: 'index-get',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function post(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'POST', path: '/post' },
    functionId: 'index-post',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function put(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'PUT', path: '/put' },
    functionId: 'index-put',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function del(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'DELETE', path: '/del' },
    functionId: 'index-del',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function patch(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'PATCH', path: '/patch' },
    functionId: 'index-patch',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function head(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'HEAD', path: '/head' },
    functionId: 'index-head',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function options(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'OPTIONS', path: '/options' },
    functionId: 'index-options',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export default function $default(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'ALL', path: '/' },
    functionId: 'index',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

`;

exports[`with requestClient generate client 1`] = `
"import { $fetch as request$0 } from '@midwayjs/fetch';
import { websocket as request$1 } from '@midwayjs/fetch';

export function get(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'GET', path: '/server/api/get' },
    functionId: 'server-api-index-get',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function custom(...args) {
  const route = {
    trigger: { type: 'Custom', isCustom: true },
    functionId: 'server-api-index-custom',
    hasMetadataInput: false,
  };
  return request$1(route, ...args);
}
"
`;

exports[`with requestClient generate multi client 1`] = `
import { $fetch as request$0 } from '@midwayjs/fetch';
import { cloud as request$1 } from 'cloud-invoker';

export function get(...args) {
  const route = {
    trigger: { type: 'HTTP', method: 'GET', path: '/server/api/get' },
    functionId: 'server-api-index-get',
    hasMetadataInput: false,
  };
  return request$0(route, ...args);
}

export function custom(...args) {
  const route = {
    trigger: { type: 'Custom', isCustom: true },
    functionId: 'server-api-index-custom',
    hasMetadataInput: false,
  };
  return request$1(route, ...args);
}

`;
